xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +136,777 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
 func (bb) Name() string {
 	return Name
 }
+
+// scUpdate wraps a subConn update to be sent to the child balancer.
+type scUpdate struct {
+	scw   *subConnWrapper
+	state balancer.SubConnState
+}
+
+type ejectionUpdate struct {
+	scw       *subConnWrapper
+	isEjected bool // true for ejected, false for unejected
+}
+
+type lbCfgUpdate struct {
+	lbCfg *LBConfig
+	// to make sure picker is updated synchronously.
+	done chan struct{}
+}
+
+type outlierDetectionBalancer struct {
+	// These fields are safe to be accessed without holding any mutex because
+	// they are synchronized in run(), which makes these field accesses happen
+	// serially.
+	//
+	// childState is the latest balancer state received from the child.
+	childState balancer.State
+	// recentPickerNoop represents whether the most recent picker sent upward to
+	// the balancer.ClientConn is a noop picker, which doesn't count RPC's. Used
+	// to suppress redundant picker updates.
+	recentPickerNoop bool
+
+	closed *grpcsync.Event
+	cc     balancer.ClientConn
+	logger *grpclog.PrefixLogger
+
+	// childMu protects the closing of the child and also guarantees updates to
+	// the child are sent synchronously (to uphold the balancer.Balancer API
+	// guarantee of synchronous calls).
+	//
+	// For example, run() could read that the child is not nil while processing
+	// SubConn updates, and then Close() could write to the the child, clearing
+	// the child, making it nil, then you try and update a cleared and already
+	// closed child, which breaks the balancer.Balancer API.
```

Comment:
How about:

```go
// childMu guards calls into child and also synchronizes reads of child in run() and the write in Close().
```

If you make sure `b.run()` exits before calling `b.child.Close()`, though, you don't need that last part of the sentence.

Reply:
Yup, waited for b.run() to exit. Changed docstring to just the first part of that sentence.

Reply:
Also removed the clearing of the child as an invariant of close, and the reading of the child (determining whether it's been cleared or not).